lib/Doctrine/DBAL/Schema/Index.php

```diff
@@ -351,7 +351,7 @@ private function hasSameColumnLengths(self $other) : bool
             return $length !== null;
         };
 
-        return array_filter($this->options['lengths'] ?? [], $filter)
-            === array_filter($other->options['lengths'] ?? [], $filter);
+        return array_map('intval', array_filter($this->options['lengths'] ?? [], $filter))
```

Comment:
I'd say `lengths` should always be `int[]`. Any way to enforce that at introspection/validation time?

Reply:
Yes, we could cast to int in the [`_getPortableTableIndexesList()`](https://github.com/doctrine/dbal/blob/2.9/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L865) method:

```php
$result[$keyName]['options']['lengths'][] = isset($tableIndex['length']) ? (int) $tableIndex['length'] : null;
```

Reply:
Probably better to `assert()` that it is integer-ish, rather than blindly applying `(int)`

Reply:
Asserting great, but you don't actually mean `assert()` should be used for this, right?

Reply:
I expect a crash, so yes, this should be a conditional with an exception.

Reply:
> Yes, we could cast to int in the [`_getPortableTableIndexesList()`](https://github.com/doctrine/dbal/blob/2.9/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L865) method:

Please do. It should be cast to the proper type as soon as it crosses the architecture boundary DB driver → DBAL. The rest of the code should only expect an `int`.

Reply:
I have adjusted the PR accordingly, however, I am not sure where to put the test. Can you point me in the right direction please?

Reply:
Possibly https://github.com/doctrine/dbal/blob/f28f3a83e29baca5d9b4ee6bebc11465053055aa/tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php ?

Reply:
I have added a test in 3d3973ead11bd3457898731aaebd03fb70b9defd.

Reply:
@morozov what do you think about https://github.com/doctrine/dbal/pull/3420#discussion_r244757515 ?

Reply:
I don't think this case is different from all others where we just cast it, especially given that the bug is not about the value not being an integer. It is an integer represented as a string which is true for most cases.

Unless @Ocramius can suggest a way to check the value, I'm fine with not having it. Should it be `preg_match()` or `filter_var()` or `ctype_digit()`? Note that the latter two will introduce implicit dependencies on a not yet used PHP extension.